kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,1180 @@
+     * @param   rEquationIds the equation id vector for the above element or condition
+     * @param   rCurrentProcessInfo the current process info
+     */
+    template <typename TContainerType>
+    void ApplyConstraints(ModelPart& rModelPart,
+                          TContainerType& rCurrentContainer,
+                          typename TContainerType::EquationIdVectorType& rEquationIds,
+                          ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY
+        // If no slave is found for this container , no need of going on
+        if (! HasSlaveNode(rCurrentContainer.GetGeometry()))
+        {
+            return;
+        }
+        DofsVectorType container_dofs;
+        typename TContainerType::EquationIdVectorType master_equation_ids;
+        rCurrentContainer.GetDofList(container_dofs, rCurrentProcessInfo);
+        IndexType slave_equation_id;
+
+        // For each node check if it is ac slave or not If it is .. we change the Transformation matrix
+        for (IndexType j = 0; j < container_dofs.size(); j++)
+        {
+            slave_equation_id = container_dofs[j]->EquationId(); // consider everything as a slave.
+            // Get the global constraint equation for this slave.
+            auto global_master_slave_constraint = mGlobalMasterSlaveRelations.find(slave_equation_id);
+            if (global_master_slave_constraint != mGlobalMasterSlaveRelations.end())
+            { // if a equation exists for this slave
+                global_master_slave_constraint->second->EquationIdVector(slave_equation_id, master_equation_ids); // get the slave and master equation ids for this slave.
+                rEquationIds.reserve(master_equation_ids.size());
+                for (auto &master_eq_id : master_equation_ids)
+                {
+                    // Add the current slaves master eq ids to the equation ids
+                    rEquationIds.push_back(master_eq_id);
+                }
+            }
+        }
+        KRATOS_CATCH("ResidualBasedBlockBuilderAndSolverWithConstraints::ApplyConstraints failed ..");
+    }
+
+    /**
+     * @brief   This function modifies the LHS and RHS of the rCurrentContainer to account for any master-slave constraints its nodes/dofs 
+     *          are carrying.
+     * @details Here cannot use the pure Geometry because, we would need the dof list from the element/geometry.
+     * @param   rModelPart The model part of the problem to solve
+     * @param   rCurrentContainer the element or condition where the rEquationIds to be modified for master-slave constraints
+     * @param   rLHSContribution the LHS contibution of the rCurrentContainer
+     * @param   rRHSContribution the RHS contibution of the rCurrentContainer
+     * @param   rEquationIds the equation id vector for the above element or condition
+     * @param   rCurrentProcessInfo the current process info
```

Comment:
My initial Idea was also something like this .. but only thing I did not like about this is that the implementation is distributed (together with the other class you mentioned above) into two or more different classes and for me it felt inconvenient to follow .. !

Reply:
I don't think it's inconvenient to have multiple classes if it helps you accomplish something. If using more classes leads to simpler or more efficient code I would not try to avoid it. You can hide the helper classes in the private section if you prefer.

Reply:
I Would suggest the following change : 

auxiliary_global_master_slave_relation.h -> constraint_builder_and_solver_helpers.h

and putting these helper classes in that file in the Internals namespace

Reply:
It's up to you, but if they are only used by the constraint builder and solver you can put it in the same file.

Reply:
just one comment on the cr_imposer: ensure that the copy constructor works correctly, since it is required by the firstprivate function of openmp